File: mobile/lib/viewmodel/article_view_model.dart
import 'package:flutter/material.dart';
import 'package:newztable/core/constants/app_constants.dart';
import 'package:newztable/core/network/api_client.dart';
import 'package:newztable/model/article_model.dart';

enum ArticleState { idle, loading, success, error }

class ArticleViewModel extends ChangeNotifier {
  ArticleState _state = ArticleState.idle;
  ArticleModel? _article;
  List<ArticleModel> _relatedArticles = [];
  String _errorMessage = '';
  bool _isBookmarked = false;

  ArticleState get state => _state;
  ArticleModel? get article => _article;
  List<ArticleModel> get relatedArticles => _relatedArticles;
  String get errorMessage => _errorMessage;
  bool get isBookmarked => _isBookmarked;

  Future<void> fetchArticle(String articleId) async {
    _setState(ArticleState.loading);

    try {
      final response = await ApiClient.get(
        '/api/articles/$articleId'
      );

      final data = response.data as Map<String, dynamic>;

      _article = ArticleModel.fromJson(
        data['article'] as Map<String, dynamic>
      );

      final List<dynamic> related = data['related'] as List<dynamic>;
      _relatedArticles = related
        .map((json) => ArticleModel.fromJson(json as Map<String, dynamic>))
        .toList();
      _setState(ArticleState.success);

    } catch(e) {
      _errorMessage = e.toString();
      _setState(ArticleState.error);
    }
  }

  Future<void> toggleBookmark(String articleId) async {
    final wasBookmarked = _isBookmarked;
    _isBookmarked = !_isBookmarked;
    notifyListeners();

    try {
      if(wasBookmarked) {
        await ApiClient.delete(
          '${AppConstants.bookmarksEndpoint}/$articleId'
        );
      } else {
        await ApiClient.post(
          AppConstants.bookmarksEndpoint,
          data: {'articleId': articleId }
        );
      }
    } catch(e) {
      _isBookmarked = wasBookmarked;
      _errorMessage = e.toString();
      notifyListeners();
    }
  }

  void setBookmarkStatus(bool isBookmarked) {
    _isBookmarked = isBookmarked;
    notifyListeners();
  }

  void _setState(ArticleState state){
    _state = state;
    notifyListeners();
  }
}

File: web/ViewModels/ArticleViewModel.cs
using web.Models;
using web.Services;

namespace web.ViewModels
{
    public enum ArticleState { Idle, Loading, Success, Error}
    public class ArticleViewModel
    {
        private readonly ApiService _apiService;

        public ArticleState State { get; private set; } = ArticleState.Idle;
        public ArticleModel? CurrentArticle { get; private set; }
        public List<ArticleModel> RelatedArticles { get; private set; } = new();
        public string ErrorMessage { get; private set; } = string.Empty;
        public bool IsBookmarked { get; private set; }

        public event Action? OnChange;

        public ArticleViewModel(ApiService apiService)
        {
            _apiService = apiService;
        }
        public async Task FetchArticleAsync(string articleId)
        {
            SetState(ArticleState.Loading);
            try
            {
                var (article, related) = await _apiService.GetArticleAsync(articleId);
                CurrentArticle = article;
                RelatedArticles = related;
                SetState(ArticleState.Success);

            } catch(Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(ArticleState.Error);
            }
        }
        public void SetArticle(ArticleModel article, bool isBookmarked)
        {
            CurrentArticle = article;
            IsBookmarked = isBookmarked;
            NotifyStateChanged();
        }

        public async Task ToggleBookmarkAsync(string articleId)
        {
            var wasBookmarked = IsBookmarked;
            IsBookmarked = !IsBookmarked;
            NotifyStateChanged();

            try
            {
                if(wasBookmarked)
                {
                    await _apiService.RemoveBookmarkAsync(articleId);
                }
                else
                {
                    await _apiService.AddBookmarkAsync(articleId);
                }
            } catch
            {
                IsBookmarked = wasBookmarked;
                NotifyStateChanged();
            }
        }
        public void SetBookmarkStatus(bool isBookmarked)
        {
            IsBookmarked = isBookmarked;
            NotifyStateChanged();
        }

        public void Reset()
        {
            CurrentArticle = null;
            RelatedArticles = new List<ArticleModel>();
            IsBookmarked = false;
            ErrorMessage = string.Empty;
            State = ArticleState.Idle;
        }

        private void SetState(ArticleState state)
        {
            State = state;
            NotifyStateChanged();
        }
        private void NotifyStateChanged() => OnChange?.Invoke();
    }
}
